### PR TITLE
[CVCUDA] PP-OCR detector preprocessor integrate CV-CUDA

### PR DESCRIPTION
**Pull request by wang-xinyu**
### PR types(PR类型)

CV-CUDA | Model

### Description

- PP-OCR detector preprocessor integrate CV-CUDA
- Optimize ProcessorManager
- Adapt PaddleClas preprocessor to use ProcessorManager register api, etc.
- Add a proc_lib member and logic in base Processor

### CODE REVIEW

**Comment by wang-xinyu** [comment]
## Test Environment
CPU: Intel(R) Xeon(R) Gold 6271C CPU @ 2.60GHz
GPU: T4
Models:
   - https://paddleocr.bj.bcebos.com/PP-OCRv3/chinese/ch_PP-OCRv3_det_infer.tar
   - https://paddleocr.bj.bcebos.com/dygraph_v2.0/ch/ch_ppocr_mobile_v2.0_cls_infer.tar
   - https://paddleocr.bj.bcebos.com/PP-OCRv3/chinese/ch_PP-OCRv3_rec_infer.tar

Model config:
```
  int cls_batch_size = 1;
  int rec_batch_size = 6;

  det_option.SetTrtInputShape("x", {1, 3, 64,64}, {1, 3, 640, 640}, {1, 3, 960, 960});
  cls_option.SetTrtInputShape("x", {1, 3, 48, 10}, {cls_batch_size, 3, 48, 320}, {cls_batch_size, 3, 48, 1024});
  rec_option.SetTrtInputShape("x", {1, 3, 48, 10}, {rec_batch_size, 3, 48, 320}, {rec_batch_size, 3, 48, 2304});
```

Test Image: https://gitee.com/paddlepaddle/PaddleOCR/raw/release/2.6/doc/imgs/12.jpg (720x1150)
Backend: TensorRT FP16
Warmup 100 rounds，tested 1000 rounds and get avg. latency.

## OpenCV
```
DBDetector preprocess: 3.212
Classifier preprocess: 0.13
Classifier preprocess: 0.131
Classifier preprocess: 0.126
Classifier preprocess: 0.075
Recognizer preprocess: 0.974
Predict Avg.: 28.3392
det boxes: [[42,413],[483,391],[484,428],[43,450]]rec text: 上海斯格威铂尔大酒店 rec score:0.979834 cls label: 0 cls score: 1.000000
det boxes: [[187,456],[399,448],[400,480],[188,488]]rec text: 打浦路15号 rec score:0.965006 cls label: 0 cls score: 1.000000
det boxes: [[23,507],[513,488],[515,529],[24,548]]rec text: 绿洲仕格维花园公寓 rec score:0.993761 cls label: 0 cls score: 1.000000
det boxes: [[74,553],[427,542],[428,571],[75,582]]rec text: 打浦路252935号 rec score:0.948193 cls label: 0 cls score: 1.000000
```
预处理总时间：4.648ms，占端到端推理时间28.34ms的16.4%

## CV-CUDA
```
DBDetector preprocess: 0.867
Classifier preprocess: 0.146
Classifier preprocess: 0.132
Classifier preprocess: 0.091
Classifier preprocess: 0.095
Recognizer preprocess: 0.98
Predict Avg.: 24.9241

det boxes: [[42,413],[483,391],[484,428],[43,450]]rec text: 上海斯格威铂尔大酒店 rec score:0.979443 cls label: 0 cls score: 1.000000
det boxes: [[183,456],[399,448],[400,480],[184,488]]rec text: 打浦路15号 rec score:0.985840 cls label: 0 cls score: 1.000000
det boxes: [[23,507],[513,488],[515,529],[24,548]]rec text: 绿洲仕格维花园公寓 rec score:0.993652 cls label: 0 cls score: 1.000000
det boxes: [[73,553],[427,542],[428,571],[74,582]]rec text: 打浦路252935号 rec score:0.921728 cls label: 0 cls score: 1.000000
```
预处理总时间：2.31ms，占端到端推理时间24.9241ms的9.3%

端到端速度提升1.14倍

File: fastdeploy/vision/ocr/ppocr/det_preprocessor.h [hunk 1]
```diff
@@ -22,43 +23,36 @@ namespace vision {
...
-  /// Set scale value for the image normalization in detection preprocess
-  void SetScale(const std::vector<float>& scale) { scale_ = scale; }
-  /// Get scale value of the image normalization in detection preprocess
-  std::vector<float> GetScale() const { return scale_; }
-
-  /// Set is_scale for the image normalization in detection preprocess
-  void SetIsScale(bool is_scale) { is_scale_ = is_scale; }
-  /// Get is_scale of the image normalization in detection preprocess
-  bool GetIsScale() const { return is_scale_; }
+  const std::vector<std::array<int, 4>>* GetBatchImgInfo() {
```
Comment (jiangjiajun) [inline fastdeploy/vision/ocr/ppocr/det_preprocessor.h:58]: 去除SetPreprocessor接口

File: fastdeploy/vision/ocr/ppocr/det_preprocessor.cc [hunk 2]
```diff
@@ -39,64 +40,61 @@ std::array<int, 4> OcrDetectorGetInfo(FDMat* img, int max_size_len) {
...
 }
-bool OcrDetectorResizeImage(FDMat* img,
-                            int resize_w,
-                            int resize_h,
-                            int max_resize_w,
-                            int max_resize_h) {
-  Resize::Run(img, resize_w, resize_h);
+
+DBDetectorPreprocessor::DBDetectorPreprocessor(const std::vector<float>& mean,
+                                               const std::vector<float>& std) {
```
Comment (jiangjiajun) [inline fastdeploy/vision/ocr/ppocr/det_preprocessor.cc]: resize_op_ = CREATE_PROCESSOR(Resize, arguments);

```
#def CREATE_PROCESSOR(op_name, arguments):
       []() {
       x = std::make_shared_ptr(#op_name)(arguments);
       register(x);
       return x;}

```

Comment (jiangjiajun) [reply fastdeploy/vision/ocr/ppocr/det_preprocessor.cc]: resize_op = CRESTE_AND_REGISTER_PROCESSOR(Resize, 1, 1)
pad_op = cxxx();

**Review by jiangjiajun — CHANGES_REQUESTED** [review]

File: fastdeploy/vision/ocr/ppocr/dbdetector.cc [hunk 3]
```diff
@@ -69,28 +71,31 @@ bool DBDetector::Predict(const cv::Mat& img,
...
+    const std::vector<cv::Mat>& images,
+    std::vector<std::vector<std::array<int, 8>>>* det_results) {
   std::vector<FDMat> fd_images = WrapMat(images);
-  std::vector<std::array<int, 4>> batch_det_img_info;
-  if (!preprocessor_.Run(&fd_images, &reused_input_tensors_, &batch_det_img_info)) {
+  if (!preprocessor_.Run(&fd_images, &reused_input_tensors_)) {
     FDERROR << "Failed to preprocess input image." << std::endl;
     return false;
   }
+  auto batch_det_img_info = preprocessor_.GetBatchImgInfo();
```
Comment (jiangjiajun) [inline fastdeploy/vision/ocr/ppocr/dbdetector.cc:82]: 为什么需要改成这种方式，上面的方式有啥问题吗？

File: fastdeploy/vision/ocr/ppocr/dbdetector.h [hunk 4]
```diff
@@ -76,6 +76,11 @@ class FASTDEPLOY_DECL DBDetector : public FastDeployModel {
     return preprocessor_;
   }
 
+  /// Set DBDetectorPreprocessor
```
Comment (jiangjiajun) [inline fastdeploy/vision/ocr/ppocr/dbdetector.h]: 先前说到这个接口可以不用了

**Review by jiangjiajun — CHANGES_REQUESTED** [review]

File: fastdeploy/vision/ocr/ppocr/dbdetector.h (hunk 4 above)
Comment (wang-xinyu) [reply fastdeploy/vision/ocr/ppocr/dbdetector.h]: 这个没删干净，其他地方包括python接口，删掉了，这里我删一下

File: fastdeploy/vision/ocr/ppocr/dbdetector.cc (hunk 3 above)
Comment (wang-xinyu) [reply fastdeploy/vision/ocr/ppocr/dbdetector.cc:82]: 之前的batch_det_img_info是通过Preprocessor::Run函数的参数传递的， 这里用了ProcessorManager的Run函数，就不能传递这个参数了

**Review by jiangjiajun — APPROVED** [review]

**Pull request merged**